### DISCUSSION TITLE
Lab: Data Transfer

### DISCUSSION THREAD

**Discussion by mbcann01** [brad-cannell/intro.r.public#30] [Labs]
Please feel free to post questions about this lab below. You are also welcome to leave comments below about the usefulness of this lab. Which parts were helpful? What could I do to improve it? What is still unclear?

**Comment by OJONESD** [comment]
Hello Dr. Cannell,

For Task 6 (importing the Chocolate txt file) I am get an error stating argument widths is missing even though I included it. I tried making a small test.txt file in notepad and it gave me the same error when I tried to import it. I attached a screenshot of my code and the error. Do you have any advice on how to fix this?

![Screenshot 2024-06-12 153815](https://github.com/brad-cannell/intro.r.public/assets/170458536/a1085d8c-4c70-49c5-a4c6-0b9cd90e9907)

Thank you for your help,
Olivia Jones

**Comment by mbcann01** [comment]
Hi Olivia,

One issue I noticed was with the function you're using (read.fwf). That is a slightly different function than the one I use in my code. I think that should help, but please let me know if I'm wrong.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by OJONESD** [reply]
Thank you so much I must have selected the wrong one when it auto-filled suggestions. It worked when I changed it to read_fwf.

Olivia

**Comment by mbcann01** [comment]
Perfect!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by Lindo-gif** [comment]
Greetings Dr Cannell,

I was wondering, what is the difference between read.csv() and read_csv() in R, and why is one often preferred over the other?
From the notes, I know that read_csv() comes from the readr package. Could it be that read.csv() is part of base R?

**Reply by mbcann01** [reply]
Hi @Lindo-gif ,

Great question! Both base R's `read.csv()` function and `readr`'s `read_csv()` function can be used to import csv files into our R session. However, there are several benefits to using `read_csv()`. [Here is a list](https://r4ds.had.co.nz/data-import.html#compared-to-base-r) of some of the major differences outlined by the creators of `read_csv()`:

> If you’ve used R before, you might wonder why we’re not using [read.csv()](https://rdrr.io/r/utils/read.table.html). There are a few good reasons to favour readr functions over the base equivalents:
> - They are typically much faster (~10x) than their base equivalents. Long running jobs have a progress bar, so you can see what’s happening. If you’re looking for raw speed, try [data.table::fread()](https://rdrr.io/pkg/data.table/man/fread.html). It doesn’t fit quite so well into the tidyverse, but it can be quite a bit faster.
> - They produce tibbles, they don’t convert character vectors to factors, use row names, or munge the column names. These are common sources of frustration with the base R functions.
> - They are more reproducible. Base R functions inherit some behaviour from your operating system and environment variables, so import code that works on your computer might not work on someone else’s.

**Reply by rabdelkhaleq** [reply]
Dr. Cannell, is this explanation similar to why write_rds() is preferred over saveRDS()?

**Reply by mbcann01** [reply]
Hi @rabdelkhaleq ,

The specific reasons may not be exactly the same, but I imagine they are similar. For me personally, an advantage of using all of the `read_` and `write_` functions from the `readr` package (as opposed to just some of them) is that I know they are all written by the same group of programmers specifically to operate in the same way. So, it ends up reducing my cognitive load. I don't have to decide which function from which package to use, I don't have to think very hard about the information I need to supply to the function, and I don't have to wonder what result to expect. They all function reliably and consistently.

**Comment by rabdelkhaleq** [comment]
Hi Dr. Cannell!

I am having issues with the haven package as well as the readxl package. I am using the latest version of RStudio. (2025.05.1+513) but when I try to install these packages and load them, I get this back: 

Warning: package ‘readxl’ was built under R version 4.3.3 
and 
Warning: package ‘haven’ was built under R version 4.3.3

Then I get long errors. I had to use different packages to complete these assignments such as (openxlsx) and then convert from xpt to csv manually...

Is there something I am missing? Should I be using a different version of R?

Also, the lab and module quiz links to r4epi are also broken :/

**Reply by mbcann01** [reply]
Hi @mbcann01 ,

Thank you for bringing the broken links to my attention! They should be fixed now. 

Additionally, I found a couple of websites that may be able to help you understand your R build version warnings:
- https://heds.nz/posts/warning-package-built-under-r-version/
- https://forum.posit.co/t/hello-im-new-to-r-the-message-below-keeps-popping-up-when-i-try-to-install-tidyverse-or-other-packages-what-is-the-solution/163262

The bottom line is that you may be able to prevent this warning by installing the packages from source (explained in the article above), but it probably isn't necessary. You are probably safe to ignore this warning and move on.
